ngraph/frontend/onnx_import/src/editor/editor.cpp

```diff
@@ -369,3 +403,118 @@ void onnx_import::ONNXModelEditor::set_input_values(
+
+    shift_node_indexes(node_indexes);
+
+    for (auto node_vector : node_indexes)
+    {
+        auto new_end = std::unique(node_vector.begin(), node_vector.end());
+        node_vector.erase(new_end, node_vector.end());
+        replace_nodes(node_vector, node_generator, new_op_name);
+    }
+    onnx_import::register_operator(new_op_name, 1, "org.openvinotoolkit.editor", node_generator);
+}
+
+void onnx_import::ONNXModelEditor::ONNXModelEditor::replace_nodes(
+    std::vector<int> node_indexes, onnx_import::Operator node_generator, std::string new_op_name)
+{
+    std::vector<std::string> inputs;
+    std::vector<std::string> outputs;
+
+    auto nodes = model().graph().node();
+
+    // 1. create a list of inputs and outputs of new custom op.
+    for (auto node_index : node_indexes)
+    {
+        inputs.insert(
+            inputs.end(), nodes[node_index].input().begin(), nodes[node_index].input().end());
+        outputs.insert(
+            outputs.end(), nodes[node_index].output().begin(), nodes[node_index].output().end());
+    }
+
+    // 2. get internal tensors via intersection of inputs and outputs.
+    std::vector<std::string> internal_tensors;
+
+    std::set<std::string> inputs_set(inputs.begin(), inputs.end());
+    std::set<std::string> outputs_set(outputs.begin(), outputs.end());
+
+    std::set_intersection(inputs_set.begin(),
+                          inputs_set.end(),
+                          outputs_set.begin(),
+                          outputs_set.end(),
+                          std::inserter(internal_tensors, internal_tensors.end()));
+
+    auto is_external_tensor = [&node_indexes, &nodes](std::string tensor_name) {
+        bool is_external = false;
+        for (int i = 0; i < nodes.size(); ++i)
+        {
+            if (std::count(node_indexes.begin(), node_indexes.end(), i))
+            {
+                continue;
+            }
+
+            is_external = is_external ||
+                          std::count(nodes[i].input().begin(), nodes[i].input().end(), tensor_name);
+        }
+        return is_external;
+    };
+
+    auto is_on_kill_list = [&internal_tensors](std::string target_name) {
+        return std::count(internal_tensors.begin(), internal_tensors.end(), target_name);
+    };
+
+    // 3. remove internal tensors from input list.
+    auto new_end = std::remove_if(inputs.begin(), inputs.end(), is_on_kill_list);
+    inputs.erase(new_end, inputs.end());
+
+    // 4. remove from internal tensors a tensor which is used outside of replaced graph
+    new_end = std::remove_if(internal_tensors.begin(), internal_tensors.end(), is_external_tensor);
+    internal_tensors.erase(new_end, internal_tensors.end());
+
+    // 5. remove internal tensors from output list.
+    new_end = std::remove_if(outputs.begin(), outputs.end(), is_on_kill_list);
+    outputs.erase(new_end, outputs.end());
+
+    // 6. outputs have to be unique
+    new_end = std::unique(outputs.begin(), outputs.end());
```

Comment:
Is it possible that they will not be unique at this point of the algorithm? If so - can you cover it with a test?

Reply:
you are correct, since output name have to be unique in onnx model and there is no duplicates in node indexes this code is unnecessary, thanks.